compiler/rustc_typeck/src/check/expr.rs

```diff
@@ -2648,6 +2648,19 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 Some((index_ty, element_ty)) => {
                     // two-phase not needed because index_ty is never mutable
                     self.demand_coerce(idx, idx_t, index_ty, None, AllowTwoPhase::No);
+                    self.select_obligations_where_possible(false, |errors| {
+                        for error in errors {
+                            match error.obligation.predicate.kind().skip_binder() {
+                                ty::PredicateKind::Trait(predicate)
+                                    if self.tcx.is_diagnostic_item(
+                                        sym::SliceIndex,
+                                        predicate.trait_ref.def_id,
+                                    ) => {}
+                                _ => continue,
+                            }
+                            error.obligation.cause.span = idx.span;
+                        }
```

Comment:
nit: Can you move this into a helper? Maybe called `point_at_index_if_possible` or something. 

Similar to https://doc.rust-lang.org/nightly/nightly-rustc/rustc_typeck/check/fn_ctxt/struct.FnCtxt.html#method.point_at_arg_instead_of_call_if_possible